tarantool/crud.py

```diff
@@ -0,0 +1,748 @@
+class Response(object):
+    def __init__(self, response):
+        """
+        TODO: write docs
+        """
+
+        if isinstance(response, dict):
+            for response_field_name in response.keys():
+                setattr(self, response_field_name, response[response_field_name])
+        elif isinstance(response, str):
+            self.str = response
+            self.err = response
+
+
+class Result(Response):
+    """
+    Contains result's fields from result variable of crud module operation.
+    """
+
+
+class Error(Response):
+    """
+    Contains error's fields from error variable of crud module operation.
+    """
+
+
+class Crud(object):
+    """
+    Contains crud module methods.
+    """
+
+    def __init__(self, con):
+        """
+        TODO: write docs
+        """
+
+        self.con = con
+        self._crud_module_init_complete = False
+
+    def _crud_module_init(self):
+        self.require_ok = self.con.eval("ok, crud = pcall(require, 'crud') return ok").data[0]
+        self.con.eval("""
+            function crud_pairs(space_name, conditions, opts) 
+                res = {} 
+                for _, v in crud.pairs(space_name, conditions, opts) do 
+                    table.insert(res, v) 
+                end 
+                return res 
+            end
+        """)
+        self.con.eval("""
+            function pcall_crud_pairs(space_name, conditions, opts) 
+                ok, res = pcall(crud_pairs, space_name, conditions, opts) 
+                return ok, res 
+            end
+        """)
+        self._crud_module_init_complete = True
+
+    def insert(self, space_name, values, opts={}):
+        """
+        TODO: write docs
+        """
+
+        if not self._crud_module_init_complete:
+            self._crud_module_init()
+        assert self.require_ok == True, 'Problem with require of crud module on the remote machine'
+        assert isinstance(space_name, str)
+        assert isinstance(values, (tuple, list))
+        assert isinstance(opts, dict)
+
+        crud_resp = self.con.call("crud.insert", space_name, values, opts)
```

Comment:
We also need to think a bit about grants. Using `eval` requires universe grants or something really close to that. Using `call`s requires dramatically less grants. If we can reduce our code to using only calls then the user which are concerned about grants hierarchy will be able to create a user with grants to call several specific functions and it will work.

Reply:
Done, now the `eval` is not used, but it may be required during implementation `pairs` (I'm working on it now)

Reply:
At the moment, it has been decided not to implement `pairs`. I suppose it's resolved